marlowe-playground-client/src/Marlowe/Linter.purs

```diff
@@ -0,0 +1,154 @@
+module Marlowe.Linter
+  ( lint
+  , State(..)
+  , Position
+  , _holes
+  , _negativeDeposits
+  , _negativePayments
+  ) where
+
+import Prelude
+import Data.Array (catMaybes, fold, foldMap, (:))
+import Data.BigInteger (BigInteger)
+import Data.Lens (Lens', over, set, view)
+import Data.Lens.Iso.Newtype (_Newtype)
+import Data.Lens.Record (prop)
+import Data.Maybe (Maybe(..))
+import Data.Newtype (class Newtype)
+import Data.Symbol (SProxy(..))
+import Data.Tuple.Nested (type (/\), (/\))
+import Marlowe.Holes (Action(..), Case(..), Contract(..), Holes, Term(..), Value(..), getHoles, insertHole)
+import Text.Parsing.StringParser (Pos)
+
+type Position
+  = { start :: Pos, end :: Pos }
+
+newtype State
+  = State
+  { holes :: Holes
+  , negativePayments :: Array Position
+  , negativeDeposits :: Array Position
+  }
+
+derive instance newtypeState :: Newtype State _
+
+instance semigroupState :: Semigroup State where
+  append (State s1) (State s2) =
+    State
+      { holes: s1.holes <> s2.holes
+      , negativePayments: s1.negativePayments <> s2.negativePayments
+      , negativeDeposits: s1.negativeDeposits <> s2.negativeDeposits
+      }
+
+instance monoidState :: Monoid State where
+  mempty =
+    State
+      { holes: mempty
+      , negativePayments: mempty
+      , negativeDeposits: mempty
+      }
+
+_holes :: Lens' State Holes
+_holes = _Newtype <<< prop (SProxy :: SProxy "holes")
+
+_negativePayments :: Lens' State (Array Position)
+_negativePayments = _Newtype <<< prop (SProxy :: SProxy "negativePayments")
+
+_negativeDeposits :: Lens' State (Array Position)
+_negativeDeposits = _Newtype <<< prop (SProxy :: SProxy "negativeDeposits")
+
+lint :: Term Contract -> State
+lint = traverseContract mempty
+
+-- | We go through a contract term collecting all warnings and holes etc so that we can display them in the editor
+traverseContract :: State -> Term Contract -> State
+traverseContract state (Term Close _ _) = state
+
+traverseContract state (Term (Pay acc payee token payment contract) start end) =
+  let
+    gatherHoles = getHoles acc <> getHoles payee <> getHoles token <> getHoles payment
+
+    newState =
+      over _holes gatherHoles
+        <<< over _negativePayments (maybeCons (negativeValue payment))
+        $ state
+  in
+    traverseContract newState contract
+
+traverseContract state (Term (If obs c1 c2) _ _) =
+  let
+    gatherHoles = getHoles obs
+
+    newState = over _holes gatherHoles state
+  in
+    traverseContract newState c1 <> traverseContract newState c2
+
+traverseContract state (Term (When cases timeout contract) start end) =
+  let
+    hs = view _holes state
+
+    (holes /\ mnds /\ contracts) = collectFromTuples (map (contractFromCase hs) cases)
+
+    nds = catMaybes mnds
+
+    newState =
+      over _negativeDeposits (append nds)
+        <<< set _holes (insertHole timeout (fold holes))
```

Comment:
Purely a purty thing, but I have to wonder if it would autoformat a bit better as:

```purescript
    newState =
      state
        # over _negativeDeposits (append nds)
        # set _holes (insertHole timeout (fold holes))
```

Unsure. Worth checking?